### ISSUE TITLE
Wrong step value in parameters sent to callback function

### ISSUE THREAD

**Issue by TimothyAlexisVass** [huggingface/diffusers#5233]
### Describe the bug

At least StableDiffusionXLPipeline and StableDiffusionXLImg2ImgPipeline sends step to callback function, starting at 1 and incrementing by 2 instead of incrementing by 1.

### Reproduction

```python
import torch

from diffusers import AutoencoderKL, StableDiffusionXLPipeline, StableDiffusionXLImg2ImgPipeline, KDPM2AncestralDiscreteScheduler

vae = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix",
    torch_dtype=torch.float16
)

base = StableDiffusionXLPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0",
    vae=vae,
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True
)

scheduler = KDPM2AncestralDiscreteScheduler.from_config(base.scheduler.config, use_karras_sigmas=True)
base.scheduler = scheduler

base.to("cuda")

def print_step(s, t, latents):
    print(s)

generator=torch.manual_seed(1111)

images = base(
    prompt="LOVE",
    num_inference_steps=10,
    generator=generator,
    callback=print_step
).images

for image in images:
    display(image)
```

### Output from print_step function:
```shell
1
3
5
7
9
...
```

### Expected output:
```
0
1
2
3
4
...
```

### System Info

Diffusers 0.21.4

### Who can help?

@PatrickSchrML @sayakpaul @williamberman

**Comment by TimothyAlexisVass** [comment]
Looking at the code, I can't understand why this would happen:

### src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
```python
        with self.progress_bar(total=num_inference_steps) as progress_bar:
            for i, t in enumerate(timesteps): # i should definitely be 1, 2, 3, 4, 5 ... here
                
                ... Can't see that i is modified in any of the lines here ...
                
                # call the callback, if provided
                if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
                    progress_bar.update()
                    if callback is not None and i % callback_steps == 0:
                        callback(i, t, latents) # callback is indeed called with i, which should definitely be 1, 2, 3, 4, 5 ...
```

**Comment by sayakpaul** [comment]
Cc: @patrickvonplaten. Could you take a look here?

**Comment by TimothyAlexisVass** [comment]
The minimal reproducible example I have provided is too minimal.
I will need to adjust it. Maybe I'll figure out why it's happening at the same time.

**Comment by TimothyAlexisVass** [comment]
OK, I'm satisfied with the minimal reproducible example and I've also tested this with 0.21.4, so I've also updated the version.

**Comment by bghira** [comment]
KDPM2AncestralDiscreteScheduler is a double order scheduler, and that seems to only ever call the callback when the equivalent of an optimization step is done. huh. that is weird.

**Comment by patrickvonplaten** [comment]
Should be fixed by #5261 - feel free to give it a try :-)
